plugins/outputs/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,322 @@
+  ## The metric timestamp will be used to decide the destination index name
+  # %Y - year (2016)
+  # %y - last two digits of year (00..99)
+  # %m - month (01..12)
+  # %d - day of month (e.g., 01)
+  # %H - hour (00..23)
+  index_name = "telegraf-%Y.%m.%d" # required.
+
+  ## Template Config
+  ## Set to true if you want telegraf to manage its index template.
+  ## If enabled it will create a recommended index template for telegraf indexes
+  manage_template = true
+  ## The template name used for telegraf indexes
+  template_name = "telegraf"
+  ## Set to true if you want telegraf to overwrite an existing template
+  overwrite_template = false
+`
+
+func (a *Elasticsearch) Connect() error {
+	if a.URLs == nil || a.IndexName == "" {
+		return fmt.Errorf("Elasticsearch urls or index_name is not defined")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), a.Timeout.Duration)
+	defer cancel()
+
+	var clientOptions []elastic.ClientOptionFunc
+
+	clientOptions = append(clientOptions,
+		elastic.SetSniff(a.EnableSniffer),
+		elastic.SetURL(a.URLs...),
+		elastic.SetHealthcheckInterval(a.HealthCheckInterval.Duration),
+	)
+
+	if a.Username != "" && a.Password != "" {
+		clientOptions = append(clientOptions,
+			elastic.SetBasicAuth(a.Username, a.Password),
+		)
+	}
+
+	if a.HealthCheckInterval.Duration == 0 {
+		clientOptions = append(clientOptions,
+			elastic.SetHealthcheck(false),
+		)
+		log.Printf("D! Elasticsearch output: disabling health check")
+	}
+
+	client, err := elastic.NewClient(clientOptions...)
+
+	if err != nil {
+		return err
+	}
+
+	// check for ES version on first node
+	esVersion, err := client.ElasticsearchVersion(a.URLs[0])
+
+	if err != nil {
+		return fmt.Errorf("Elasticsearch version check failed: %s", err)
+	}
+
+	// quit if ES version is not supported
+	i, err := strconv.Atoi(strings.Split(esVersion, ".")[0])
+	if err != nil || i < 5 {
+		return fmt.Errorf("Elasticsearch version not supported: %s", esVersion)
+	}
+
+	log.Println("I! Elasticsearch version: " + esVersion)
+
+	a.Client = client
+
+	if a.ManageTemplate {
+		err := a.manageTemplate(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (a *Elasticsearch) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 {
+		return nil
+	}
+
+	bulkRequest := a.Client.Bulk()
+
+	for _, metric := range metrics {
+		var name = metric.Name()
+
+		// index name has to be re-evaluated each time for telegraf
+		// to send the metric to the correct time-based index
+		indexName := a.GetIndexName(a.IndexName, metric.Time())
+
+		m := make(map[string]interface{})
+		mfields := make(map[string]interface{})
+
+		for k, v := range metric.Fields() {
+			switch x := v.(type) {
+			// Truncate values too big/small for Elasticsearch to not complain when creating a dynamic field mapping
+			case float64:
+				if (x > 0 && x > math.MaxFloat32) || (x < 0 && x < math.SmallestNonzeroFloat32) {
```

Comment:
The 0 comparison I added because I saw truncation occurring when values were 0 (or -0, not sure), I will double check this later today.

Isn't "float32" in go the same as "float" in ES? Because it seems ES can handle fine these. I will check again if ES dynamic field mapping does not blow with MaxFloat32 and SmallestNonzeroFloat32 (and maybe add a new test).